src/bin/pg_autoctl/keeper.c

```diff
@@ -113,6 +113,53 @@ keeper_update_state(Keeper *keeper, int node_id, int group_id,
 }
 
 
+/*
+ * keeper_should_ensure_current_state returns true when pg_autoctl should
+ * ensure that Postgres is running, or not running, depending on the current
+ * FSM state, before calling the transition function to the next state.
+ *
+ * At the moment, the only cases when we DON'T want to ensure the current state
+ * are when either the current state or the goal state are one of the following:
+ *
+ *  - DRAINING
+ *  - DEMOTED
+ *  - DEMOTE TIMEOUT
+ *
+ * That's because we would then stop Postgres first when going from DEMOTED to
+ * SINGLE, or ensure Postgres is running when going from PRIMARY to DEMOTED.
+ * This last example is a split-brain hazard, too.
+ */
+bool
+keeper_should_ensure_current_state_before_transition(Keeper *keeper)
+{
+	KeeperStateData *keeperState = &(keeper->state);
+
+	if (keeperState->assigned_role != keeperState->current_role)
+	{
+		if (keeperState->assigned_role == DEMOTED_STATE
+			|| keeperState->assigned_role == DEMOTE_TIMEOUT_STATE
+			|| keeperState->assigned_role == DEMOTED_STATE)
+		{
+			/* don't ensure Postgres is running before shutting it down */
+			return false;
+		}
+
+		if (keeperState->current_role == DEMOTED_STATE
```

Comment:
what if assigned_role != <ANY_DEMOTED_STATE> but current role is ?

Reply:
node is currently demoted, but assigned state is something else

Reply:
Well then as the comment says: _ don't ensure Postgres is down before starting it again_. What am I missing here?

Reply:
it made sense when you state it that way, thanks